elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/metric/ComputedMetric.java

```diff
@@ -0,0 +1,209 @@
+        this.description = metaData == null || metaData.description().isEmpty() ? metricField : metaData.description();
+        this.dataType = fieldType;
+        this.aggregations = MetricUtils.extractAggregations(metricField, cls, entityDictionary);
+
+        if (aggregations.isEmpty()) {
+            String message = String.format("'%s' in '%s' has no aggregation.", metricField, cls.getCanonicalName());
+            log.error(message);
+            throw new IllegalStateException(message);
+        }
+
+        this.metricExpression = expandMetricExpression(metricField, cls, new HashSet<>(), entityDictionary);
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) { return true; }
+        if (o == null || getClass() != o.getClass()) { return false; }
+        final ComputedMetric that = (ComputedMetric) o;
+        return getName().equals(that.getName())
+                && getLongName().equals(that.getLongName())
+                && getDescription().equals(that.getDescription())
+                && getDataType().equals(that.getDataType())
+                && getAggregations().equals(that.getAggregations())
+                && getMetricExpression().equals(that.getMetricExpression());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(
+                getName(),
+                getLongName(),
+                getDescription(),
+                getDataType(),
+                getAggregations(),
+                getMetricExpression()
+        );
+    }
+
+    @Override
+    public String toString() {
+        return new StringJoiner(", ", ComputedMetric.class.getSimpleName() + "[", "]")
+                .add("name='" + name + "'")
+                .add("longName='" + longName + "'")
+                .add("description='" + description + "'")
+                .add("dataType=" + dataType)
+                .add("aggregations=" + aggregations)
+                .add("metricExpression='" + metricExpression + "'")
+                .toString();
+    }
+
+    /**
+     * Returns the <b>expanded</b> JPQL metric metricExpression, with {@link ComputedMetric#TABLE_PREFIX table prefix
+     * placeholder} attached, for computing a specified entity field.
+     * <p>
+     * See {@link ComputedMetric#TABLE_PREFIX} for more details.
+     *
+     * @param expandingField  The entity field
+     * @param containingCls  The entity that contains the {@code expandingField}
+     * @param visited  A set of already-visited metricExpression variables, which should be skipped during expansion.
+     *
+     * @return a JPQL metricExpression with placeholder
+     */
+    private static String expandMetricExpression(
+            String expandingField,
+            Class<?> containingCls,
+            Set<String> visited,
+            EntityDictionary entityDictionary
+    ) {
+        if (MetricUtils.isBaseMetric(expandingField, containingCls, entityDictionary)) {
```

Comment:
We can determine if a metric is a base metric based on whether or not it has any supported aggregations.